### PR TITLE
[DOCS] Reformat index API.

### PR DESCRIPTION
**Pull request by debadair**
This PR updates the doc index API to align with the new [API reference template](https://github.com/elastic/docs/blob/master/shared/api-ref-ex.asciidoc).

Relates to elastic/docs#937 and #45365

A few questions wrt applying the API Ref template and terminology:

- How are we handling cases where particular path params are mutually exclusive? For example,  the typical usage is `PUT /<index>/_doc/<_id>`, but you can specify `PUT /<index>/_create/<_id>` to get the _put if absent_ behavior. 
- Related, how do we talk about `_whatever` params--as "endpoints", paths, something else?
- How do we want to refer to the doc `_id`? Always as `_id`, ID, doc ID, doc `_id`?

### CODE REVIEW

**Comment by jrodewig** [comment]
> How are we handling cases where particular path params are mutually exclusive? For example, the typical usage is PUT /<index>/_doc/<_id>, but you can specify PUT /<index>/_create/<_id> to get the put if absent behavior.

I would document both endpoints in the `Request` section. You can note mutually exclusive path params in the param definitions. You can also note this in the long-form [`Description` section](https://github.com/elastic/docs/blame/master/shared/api-ref-ex.asciidoc#L40).

> Related, how do we talk about _whatever params--as "endpoints", paths, something else?

If you need to refer to them, "resource" would be the most common term. "Endpoint" usually refers to the entire URL.

> How do we want to refer to the doc _id? Always as _id, ID, doc ID, doc _id?

I think `_id` is appropriate here as that's also the meta field used to return document IDs. We can use document ID if we're referring to the value, not the field or param.

File: docs/reference/docs/index_.asciidoc [hunk 1]
```diff
@@ -1,11 +1,124 @@
...
+`POST /<index>/_doc/`
+
+[[docs-index-api-path-params]]
+==== {api-path-parms-title}
+
+`<index>`::
+(Required, string) Name of the target index. By default, the index is created
+automatically if it doesn't exist. For more information, see <<index-creation>>.
+
+`_doc`::
```
Comment (jrodewig) [inline docs/reference/docs/index_.asciidoc]: I don't think we should document `_doc` as it's no longer a changeable param. We can just remove this definition.

File: docs/reference/docs/index_.asciidoc [hunk 2]
```diff
@@ -39,22 +152,9 @@ The result of the above index operation is:
...
-
-NOTE:   Replica shards may not all be started when an indexing operation successfully returns (by default, only the
-        primary is required, but this behavior can be <<index-wait-for-active-shards,changed>>). In that case,
-        `total` will be equal to the total shards based on the `number_of_replicas` setting and `successful` will be
-        equal to the number of shards started (primary plus replicas). If there were no failures, the `failed` will be 0.
-
 [float]
 [[index-creation]]
-==== Automatic Index Creation
+==== Automatic index creation
```
Comment (jrodewig) [inline docs/reference/docs/index_.asciidoc]: I would consider moving this and the following sections into a long-form [`Description` section](https://github.com/elastic/docs/blame/master/shared/api-ref-ex.asciidoc#L39).

Some of these sections seem to be explaining functionality vs. providing examples.

File: docs/reference/rest-api/common-parms.asciidoc [hunk 3]
```diff
@@ -70,6 +70,56 @@ tag::cat-v[]
 to `false`.
 end::cat-v[]
 
+tag::doc-pipeline[]
+`pipeline`::
+(Optional, string) The ID of the pipeline to use to preprocess incoming
```
Comment (jrodewig) [inline docs/reference/rest-api/common-parms.asciidoc]:
```suggestion
(Optional, string) ID of the pipeline to use to preprocess incoming
```

File: docs/reference/rest-api/common-parms.asciidoc [hunk 4]
```diff
@@ -70,6 +70,56 @@ tag::cat-v[]
...
+documents.
+end::doc-pipeline[]
+
+tag::doc-refresh[]
+`refresh`::
+(Optional, enum) If `true` then refresh the affected shards to make this
```
Comment (jrodewig) [inline docs/reference/rest-api/common-parms.asciidoc]: Current wording could imply the user is expected to refresh the shards. We may want to explicitly state ES will refresh if `true`.

**Review by jrodewig — APPROVED** [review]
LGTM. Left a few comments, but nothing blocking.

File: docs/reference/rest-api/common-parms.asciidoc (hunk 4 above)
Comment (debadair) [reply docs/reference/rest-api/common-parms.asciidoc]: Fixed

File: docs/reference/rest-api/common-parms.asciidoc (hunk 3 above)
Comment (debadair) [reply docs/reference/rest-api/common-parms.asciidoc]: Fixed.

File: docs/reference/docs/index_.asciidoc (hunk 2 above)
Comment (debadair) [reply docs/reference/docs/index_.asciidoc]: Done.

File: docs/reference/docs/index_.asciidoc (hunk 1 above)
Comment (debadair) [reply docs/reference/docs/index_.asciidoc]: Removed & addressed _doc vs _create in the detailed description section.

**Comment by debadair** [comment]
Fixed the broken link with: https://github.com/elastic/elasticsearch-hadoop/pull/1331

**Pull request merged**